File: bit-manipulation/2683.js


/**
 * @param {number[]} derived
 * @return {boolean}
 */
var doesValidArrayExist = function (derived) {
    const n = derived.length;

    const original = new Array(n).fill(0);
    original[0] = 0;
    for (let i = 0; i < n - 1; i++) {
        original[i + 1] = original[i] ^ derived[i];
    }
    if ((original[n - 1] ^ original[0]) === derived[n - 1]) {
        return true;
    }

    original[0] = 1;
    for (let i = 0; i < n - 1; i++) {
        original[i + 1] = original[i] ^ derived[i];
    }
    if ((original[n - 1] ^ original[0]) === derived[n - 1]) {
        return true;
    }
    return false;
};



var doesValidArrayExist = function (derived) {
    let xor = 0
    for (let x of derived) {
        xor = xor ^ x
    }
    return xor === 0

};

File: bit-manipulation/2683.py
# Samsung, Apple, Microsoft, Netflix
# Tc = O(n)
# Sc = O(n)

class Solution:
    def doesValidArrayExist(self, derived: List[int]) -> bool:
        n = len(derived)
        original = [0] * n
        original[0] = 0
        for i in range(n-1):
            original[i+1] = original[i] ^ derived[i]
        if original[n-1] ^ original[0] == derived[n-1]:
            return True
        original[0] = 1
        for i in range(n-1):
            original[i+1] = original[i] ^ derived[i]
        if original[n-1] ^ original[0] == derived[n-1]:
            return True
        return False

# Approach-2
# Tc = O(n)
# Sc = O(1)
class Solution:
    def doesValidArrayExist(self, derived: List[int]) -> bool:
        XOR = 0
        for x in derived:
            XOR = XOR ^ x
        return XOR == 0